### PR TITLE
prov/gni: refactor of gni nic code

### PR DESCRIPTION
**Pull request by hppritcha**
Refactor gni_nic code into a separate
set of files specific to gni nic implementation.
Add tx desc list management.

@sungeunchoi 
@bturrubiates 
@ztiffany 

Signed-off-by: Howard Pritchard howardp@lanl.gov

### CODE REVIEW

**Comment by bturrubiates** [comment]
I think gnix_nic.h should belong under include, not src.

File: prov/gni/src/gnix_nic.h [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+ */
+
+extern uint32_t gnix_def_max_nics_per_ptag;
+
+/*
+ * prototypes
+ */
+
+int gnix_nic_tx_freelist_init(struct gnix_nic *nic, int n_descs);
+int gnix_nic_free(struct gnix_nic *nic);
```
Comment (bturrubiates) [inline prov/gni/src/gnix_nic.h]: Are these internal use only? Should these be named _gnix_nic*?

File: prov/gni/src/gnix_nic.c [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+		pthread_mutex_lock(&gnix_nic_list_lock);
+
+		gnix_list_del_init(&nic->gnix_nic_list);
+		--gnix_nics_per_ptag[nic->ptag];
+		gnix_list_del_init(&nic->list);
+
+		pthread_mutex_unlock(&gnix_nic_list_lock);
+		free(nic);
+	}
+
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_nic.c:137]: Could `assert(v>=0)` here.

**Comment by sungeunchoi** [comment]
I think I might be confused.  Where is `gnix_id_counter` declared now?

**Comment by bturrubiates** [comment]
@sungeunchoi not sure why but before it was declared both in gnix_cm_nic.c and gnix_ep.c. The gnix_cm_nic.c declaration wasn't touched so it's still there, but doesn't show up on the github diff.

File: prov/gni/src/gnix_nic.h (hunk 1 above)
Comment (hppritcha) [inline prov/gni/src/gnix_nic.h]: they are internal only, will add the _.

**Comment by sungeunchoi** [comment]
Thx @bturrubiates 

Good to go :+1:

File: prov/gni/src/gnix_nic.c (hunk 2 above)
Comment (hppritcha) [inline prov/gni/src/gnix_nic.c:137]: done

**Comment by hppritcha** [comment]
`gnix_id_counter` moved to gnix_cm_nic.c
Its now being used both for the gnix_cm_nic cdm_id's and gnix_nic cdm_ids.  I needed to do this because we need distinct cdm_id's for potentially multiple RDM EP's per process.  Can't just use getpid.

**Comment by hppritcha** [comment]
looks like I need to rebase

**Pull request merged**